vendor/bat-native-ledger/src/bat/ledger/internal/core/future_join.h

```diff
@@ -0,0 +1,157 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_FUTURE_JOIN_H_
+#define BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_FUTURE_JOIN_H_
+
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "base/memory/ref_counted.h"
+#include "base/memory/scoped_refptr.h"
+#include "bat/ledger/internal/core/future.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace ledger {
+
+template <typename... Args>
+class FutureJoin : public base::RefCounted<FutureJoin<Args...>> {
+  using Promise = Promise<std::tuple<Args...>>;
+
+ public:
+  explicit FutureJoin(Promise promise) : promise_(std::move(promise)) {}
+
+  void AddFutures(Future<Args>... futures) {
+    DCHECK_EQ(remaining_, sizeof...(Args));
+    DCHECK(!started_);
+    if (!started_) {
+      started_ = true;
+      AddHandlers<0>(std::move(futures)...);
+    }
+  }
+
+ private:
+  friend class base::RefCounted<FutureJoin>;
+
+  ~FutureJoin() = default;
+
+  template <size_t Index, typename T, typename... Rest>
+  void AddHandlers(Future<T> future, Rest... rest) {
+    future.Then(base::BindOnce(&FutureJoin::OnComplete<Index, T>, this));
+    AddHandlers<Index + 1>(std::move(rest)...);
+  }
+
+  template <size_t Index>
+  void AddHandlers() {}
```

Comment:
There are more convenient ways to do this pre-C++17, but as of C++17, you can further simplify it with a fold expression:
```cpp
template <std::size_t... Is, typename... Fs>
void AddHandlers(std::index_sequence<Is...>, Fs... fs) {
  (..., fs.Then(base::BindOnce(
            &FutureJoin::OnComplete<Is, typename Fs::ValueType>, this)));
}
```
Usage:
```cpp
AddHandlers(std::make_index_sequence<sizeof...(futures)>{},
            std::move(futures)...);
```

Reply:
Nice - done!